GitHub and google authentication

wazzuper opened the pull request:
- GitHub authentication
- Google authentication
- Modified devise views

atatarenko commented on app/views/devise/registrations/new.html.slim:
```diff
@@ -15,4 +15,7 @@
         = f.password_field :password_confirmation, class: 'form-control'
       = f.submit t('.sign_up', default: 'Sign up'), class: 'btn btn-primary'
+      hr
+      = link_to 'Sign up with GitHub', user_github_omniauth_authorize_path, class: 'btn btn-dark'
```
Don't forget about localization. It's vital thing.

atatarenko commented on app/models/user.rb:
```diff
@@ -2,6 +2,25 @@ class User < ApplicationRecord
+         :omniauthable, :confirmable
+
+  def self.from_omniauth(auth)
+    user = User.where(email: auth.info.email).first
```
JFYI, that's a way to get a single record by some field
```rb
User.find_by(email: auth.info.email)
```

atatarenko commented on app/models/user.rb:
```diff
@@ -2,6 +2,25 @@ class User < ApplicationRecord
+    user = User.where(email: auth.info.email).first
+
+    if user
+      return user
```
Check [this](https://github.com/bbatsov/ruby-style-guide#no-explicit-return) out. You can add a rubocop rule that thing.

atatarenko commented on app/models/user.rb:
```diff
@@ -2,6 +2,25 @@ class User < ApplicationRecord
+    else
+      where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+        user.email = auth.info.email
+        user.password = Devise.friendly_token[0,20]
```
Space after comma `Devise.friendly_token[0, 20]`.
Let's add a rubocop rule

atatarenko commented on app/models/user.rb:
```diff
@@ -2,6 +2,25 @@ class User < ApplicationRecord
+        user.password = Devise.friendly_token[0,20]
+        user.uid = auth.uid
+        user.provider = auth.provider
+        # If you are using confirmable and the provider(s) you use validate emails, 
```
Let's not left useless comments at the code.

atatarenko commented on app/models/user.rb:
```diff
@@ -2,6 +2,25 @@ class User < ApplicationRecord
+      end
+    end
+
+  end
```
After all I can't get this trick:
```ruby
user = User.where(email: auth.info.email).first
if user
  user
else
  where(provider: auth.provider, uid: auth.uid) do
    ...
  end
end
```
Why not just merely leave the `where(provider: auth.provider, uid: auth.uid)` part?

atatarenko commented on app/controllers/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      redirect_to new_user_registration_url
+    end
+  end
+end
```
Let's get rid of using a local variable in the actions and dry them a bit.
```ruby
private

def user
  @user ||= User.from_omniauth(request.env['omniauth.auth'])  
end
```
Then you can go with `user` property for both actions

wazzuper commented:
Everything is ok on the local server, I can sign up easily with GitHub and Google, but on heroku I get this error: ```NoMethodError (undefined method `persisted?' for #<ActiveRecord::Relation []>)```

Gist file with full log of this error https://gist.github.com/wazzuper/b0e80a642a69d3ba6678d07f149253fa

atatarenko commented:
@wazzuper oh, that's because of me. I've made a mistake when suggested `where(provider: auth.provider, uid: auth.uid)` thing.

atatarenko commented on app/models/user.rb:
```diff
@@ -1,5 +1,21 @@
+  def self.from_omniauth(auth)
+    user = User.find_by(email: auth.info.email)
+
+    if user
```
I don't really like this place cause our models should be thin. Let's move that logic to `OmniAuthService`. Put it there `app/services/omni_auth_service.rb`
```rb
class OmniAuthService
  def initialize(auth_params)
    @auth_params = auth_params
  end

  def user
    where(provider: provider, uid: uid).find_or_create do |user|
      user.assign_attributes(email: email, password: password)
      user.skip_confirmation!
    end
  end

  private

  attr_reader :auth_params
  delegate :provider, :uid, to: :auth_params

  def email
    auth_params.info.email
  end

  def password
    Devise.friendly_token[0, 20]
  end
end
```

wazzuper replied:
If I get it right, code in user.rb should look like this:
```ruby
  def self.from_omniauth(auth)
    user = User.find_by(email: auth.info.email)

    if user
      user
    else
      OmniAuthService.new(auth).user
    end
  end
```

But with those changes in user.rb, omni_auth_service.rb and controller I get an error
```undefined method `where' for #<OmniAuthService:0x00007ff409cdd1d8>```
```ruby
def user
    where(provider: provider, uid: uid).find_or_create do |user|
      user.assign_attributes(email: email, password: password)
      user.skip_confirmation!
    end
```

atatarenko replied:
Nop, let's remove `self.from_omniauth` from `User` at all.

And change `where(provider: provider, uid: uid)` to `User.where(provider: provider, uid: uid)`.

This concept I wrote, I'm sure it's 100% workable. It's just a draft of an idea.

atatarenko commented on app/controllers/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  private
+
+  def user
+    @user ||= User.from_omniauth(request.env['omniauth.auth'])
```
Take a look at my suggestion about `OmniAuthService`. Let's use it here:
```ruby
@user ||= OmniAuthService.new(request.env['omniauth.auth']).user
```

atatarenko commented:
Note: I'm not able to login via `Google` or `Github` using heroku. Could you redeploy it?

wazzuper commented:
It's finally working!
![screenshot from 2018-03-02 12-12-20](https://user-images.githubusercontent.com/17517020/36887444-15dd76a8-1e13-11e8-9b65-9854ae06f5ed.png)

atatarenko commented on reek.yml:
```diff
@@ -1,8 +1,20 @@
+  enabled: false
+
+UtilityFunction:
+  enabled: false
```
Hm, looks like the code was changed after the review, so you should re-enable these disabled metrics.
I just took a look at `DuplicateMethodCall` and `InstanceVariableAssumption` - these guys shouldn't fail anymore.

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,25 @@
+  def password
+    Devise.friendly_token[0, 20]
+  end
+end
```
Could you write a spec for this service?

atatarenko commented on app/controllers/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  def user
+    @user ||= OmniAuthService.new(request.env['omniauth.auth']).user
+  end
+end
```
Could you write a spec for this controller?

atatarenko commented on app/controllers/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,29 @@
+      sign_in_and_redirect @user, event: :authentication
+      flash[:notice] = I18n.t('devise.omniauth_callbacks.success', kind: 'GitHub')
+    else
+      session['devise.github_data'] = request.env['omniauth.auth']
```
Have you tried to remove this line and see what happen?

atatarenko commented:
@wazzuper I haven't still been able to sign-in within Github or Google

wazzuper commented:
- Modified reek.yml config
- Fixed log in/sign up with Github or Google
- Added Identity model to store provider and uid fields
- Modified omni_auth_service.rb
- Will start writing tests soon

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,45 @@
+  delegate :provider, :uid, to: :auth_params
+
+  def user
+    existing_provider = Identity.where(provider: provider, uid: uid).first
```
`where(...).first` statement can always be replaced with `find_by(...)`

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def user
+    existing_provider = Identity.where(provider: provider, uid: uid).first
+    unless existing_provider.eql?(nil)
```
`existing_provider.eql?(nil)` -> `existing_provider.nil?`

wazzuper replied:
Reek warning:
https://github.com/troessner/reek/blob/master/docs/Nil-Check.md

atatarenko replied:
:(
Try this:
```ruby
def user
  Identity.find_by!(provider: provider, uid: uid).user
rescue ActiveRecord::RecordNotFound
  create_or_return_user
end
```

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def user
+    existing_provider = Identity.where(provider: provider, uid: uid).first
+    unless existing_provider.eql?(nil)
```
If you have `else` branch in a condition statement, don't use `unless`, be positive. Consider:
```ruby
if existing_provider.nil?
  create_or_return_user(provider, uid)
else
  existing_provider.user
end
```

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,45 @@
+      user.assign_attributes(password: password)
+      user.skip_confirmation!
+    end
+    Identity.create!(
```
Do we really need that `Identity` entity? It seems a bit overkilling.

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,45 @@
+
+  def get_user
+    user
+  end
```
Can't understand your intention at that place. Why do we need this?
```ruby
def get_user
  user
end
```

wazzuper replied:
Just wanted to put ```user``` method to the private section, I don't know why, maybe to be more safety?) But yes, this is stupid - removed

atatarenko commented on app/controllers/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,29 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
Refactoring suggestion
```ruby
class OmniauthCallbacksController < Devise::OmniauthCallbacksController
  def github
    handle_oauth_callback
  end

  def google_oauth2
    handle_oauth_callback
  end

  private

  def handle_oauth_callback
    if user.persisted?
      sign_in_and_redirect user, event: :authentication
      flash[:notice] = I18n.t('devise.omniauth_callbacks.success', kind: provider)
    else
      redirect_to new_user_registration_url
    end
  end

  def provider
    user.identity.provider.humanize
  end

  def user
    @user ||= OmniAuthService.new(request.env['omniauth.auth']).user
  end
end
```

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,41 @@
+
+  delegate :provider, :uid, to: :auth_params
+
+  def create_or_return_user(provider, uid)
```
You don't need to pass `(provider, uid)` here because they are available all across the class. The line should be just `def create_or_return_user`.

`find_or_create_user` sounds more Rails-ish.

atatarenko commented on app/services/omni_auth_service.rb:
```diff
@@ -0,0 +1,41 @@
+    user.identities.create!(
+      provider: provider,
+      uid: uid,
+      user_id: user.id
```
Is it necessary? It looks like rails can automagicaly pass the user id to a creating identity.

atatarenko commented on spec/support/controller_macros.rb:
```diff
@@ -0,0 +1,9 @@
+module ControllerMacros
```
Why do we need this file? What does it do?

wazzuper replied:
Just in case, for the future tests.
Deleted! I will add it later, if necessary

atatarenko commented on spec/services/omni_auth_service_spec.rb:
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe OmniAuthService do
```
The structure of the service spec should be very close to the example below. First `describe` is a service class. Next goes the `describe`s of service public methods. Next goes the `context`s (conditions).
```ruby
RSpec.describe OmniAuthService do
  describe '#user' do
    context 'when ...' do
    end

    context 'when ...' do
    end
  end

  describe '#another_method' do
     ...
  end
end
```

atatarenko commented on spec/services/omni_auth_service_spec.rb:
```diff
@@ -0,0 +1,20 @@
+  let(:github_auth) { github_auth_hash }
+  let(:google_auth) { google_auth_hash }
+
+  describe '#user for github' do
```
Could you add more conditions to this test? Some uncovered cases:
1) New identity, existing user
2) New identity, new user
3) Existing identity
4) And so on...

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,101 @@
+RSpec.describe OmniauthCallbacksController, type: :controller do
+  describe '#github' do
+    context 'when github email doesn\'t exist' do
+      before(:each) do
```
There is no need to write `before(:each) do` explicitly, just `before do`.

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,101 @@
+        expect(authentication).to_not be_nil
+      end
+
+      it 'redirect to home page' do
```
`it 'redirects to home page'`

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,101 @@
+
+      it 'expects omniauth.auth to be be_truthy' do
+        expect(request.env['omniauth.auth']).to be_truthy
+      end
```
Try not to duplicate the description by the code. If you have no idea how to avoid it, you can just go with:
```ruby
it { expect(request.env['omniauth.auth']).to be_truthy }
```

atatarenko commented on spec/support/omniauth_macros.rb:
```diff
@@ -0,0 +1,45 @@
+module OmniauthMacros
```
This module seems a bit weird to me. First of all, this name, we don't use the word `macros` in the ruby world, `helper` is better. Secondary, I don't like the side effect. When I call `#github_auth_hash`, I am not waiting for initialization of `OmniAuth.config.mock_auth[:github]`. Finally, I don't see a reason to include it in each spec file. Looks like we need it only in `spec/services/omni_auth_service_spec.rb` so I propose to get rid of this file.

atatarenko replied:
In `spec/services/omni_auth_service_spec.rb`
```ruby
let(:omni_auth_hash) do
  OmniAuth::AuthHash.new({
    provider: 'provider',
    uid: '12345',
    info: { email: 'user@email.com' }
  })
end
```

wazzuper replied:
I'm new to this OmniAuth testing, so, used some guides from the Internet.

atatarenko commented on spec/services/omni_auth_service_spec.rb:
```diff
@@ -0,0 +1,73 @@
+      end
+    end
+  end
+end
```
You don't have to test it for `google` and `github` both, this is general purpose service and it works with every provider so it's okay to test the business logic once.

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,95 @@
+  end
+end
+
+def github_omniauth
```
This is not a good thing to use pure methods in specs. Code directly in `before` blocks if needed.

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+RSpec.describe OmniauthCallbacksController, type: :controller do
+  describe '#github' do
+    context 'when github email doesn\'t exist' do
```
```ruby
context 'when github email does not exist' do
```

atatarenko commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -0,0 +1,95 @@
+      before do
+        github_omniauth
+        get :github
+        @user = User.find_by(email: 'user@email.com')
```
Let's not use instance variables in specs, consider:
```ruby
let(:added_user) { User.find_by(email: 'user@email.com') } 
```
